### PR TITLE
Modernize Gapminders

### PR DESCRIPTION
**Pull request by jtao1**
# Modernizing an example checklist

## Preliminary checks

- [x] Look for [open PRs](https://github.com/holoviz-topics/examples/pulls) and [issues](https://github.com/holoviz-topics/examples/issues) that reference the project you are updating. It is possible previous unmerged work in PR could be re-used to modernize the project. Comment on these PRs and issues when appropriate, hopefully we should be able to close some of them after your modernizing work.

## Change ‘anaconda-project.yml’ to use the latest workable version of packages

- [x] Pin python=3.11
- [x] Remove the upper pin (e.g. `hvplot<0.9` to `hvplot`, `panel>=0.12,<1.0` to `panel>=0.12`) of all other dependencies. Removing the upper pins of dependencies could necessitate code revisions in the notebooks to address any errors encountered in the updated environment. Should complexities or extensive time requirements arise, document issues for team discussion on whether to re-pin specific packages or explore other solutions.
- [x] Add/update the lower pin of all other dependencies (e.g. `hvplot` to `hvplot>=0.9.2`, `hvplot>=0.8` to `hvplot>=0.9.2`). Usually, the new/updated lower pin of a dependency will be the version resolved after `anaconda prepare` has been run. Execute `!conda list` in a notebook, or `anaconda run conda list` in the terminal, to display the version of each dependency installed in the environment. Adjusting the lower pin helps ensure that the locks produced for each platform (linux-64, win-64, osx-64, osx-arm64) rely on the tested dependencies and not on some older versions.
- [x] If one of the channels include conda-forge or pyviz, ask Maxime if it can be removed

## Plot API updates (discussed on a per-example basis)

- [x] Generally, try to replace HoloViews usage with hvPlot. At a certain point of complexity, such as with the use of ‘.select’, it might be better to stick with HoloViews. Additional examples of ‘complexity boundaries’ should be documented in [this document](https://docs.google.com/document/d/1LYkChbVZzqq5S1S4Mkj1rM3PNPJDDGriks8YTc4EQc4/edit?usp=sharing).
- [x] Almost always, try to replace the use of `datashade` with `rasterize` (read [this page](https://holoviews.org/user_guide/Large_Data.html)). Essentially, `rasterize` allows Bokeh to handle the colormapping instead of Datashader.

## Interactivity API updates (discussed on a per-example basis)

- [x] Remove all `pn.interact` usage
- [x] Avoid `.param.watch()` usage. This is [pretty low-level and verbose approach](https://param.holoviz.org/user_guide/Dependencies_and_Watchers.html#watchers) and should not be used in Examples unless required, or an Example is specifically trying to demo its usage in an advanced workflow.
- [x] Prefer using `pn.bind()`. Read [this page](https://panel.holoviz.org/explanation/api/reactive.html) for explanation.
- [x] For apps built using a class approach, when they create a `view()` method and call it directly, update the class by inheriting from `pn.viewable.Viewer` and replace `view()` by `__panel__()`. Here is an [example](https://panel.holoviz.org/gallery/nyc_deckgl.html). 

## Panel App updates (discussed on a per-example basis)

- [x] If the project doesn’t at any point create a Panel app at all, consider creating one. It can be as simple as wrapping a plot in `pn.Column`, or more complicated to incorporate widgets, etc. Make the final app  `.servable()`.
- [x] If the project creates an app in a notebook but doesn’t deploy it (i.e. there is no `command: dashboard` declaration in the `anaconda-project.yml` file), try adding it.
- [x] If the project already deploys an app but doesn’t wrap it in a nice template, consider wrapping it in a [template](https://panel.holoviz.org/reference/index.html#templates).
- [x] If the project deploys an app wrapped in a template, customize the template a little so all the apps don’t look similar (e.g. change the header background color). This doesn’t need to be discussed.
- [x] Comment start If you are building the application in a single cell, you can construct a template explicitly, like `template = pn.template.BootstrampTemplate`, but if building up an app across multiple cells, it is probably cleaner to declare the template at the top with `pn.extension(template='bootstrap')`. See [how to guide on setting a template](https://panel.holoviz.org/how_to/templates/template_set.html).

## General code quality updates

- [x] If the notebook disables warnings (e.g. with `warnings.simplefilter(‘ignore’)` somewhere at the start of the notebook, remove this line. Try to update the code to remove the warnings, if any. If updating the code to remove the warnings is taking significant amount of time and effort, bring it up for discussion and we may decide to disable warnings again.

## Text content

- [x] Edit the text content anywhere and everywhere that it can be improved for clarity.
- [x] Check the links are valid, and update old links (e.g. http -> https, xyz.pyviz.org -> xyz.holoviz.org)
- [x] Remove instructions to install packages inside an example

## Visual appearance - Example

- [x] Check that the titles/headings make sense and are succinct.
- [x] Check that the text content blocks are easily readable; revise into additional paragraphs if needed.
- [x] Check that the code blocks are easily readable; revise as needed. (e.g. add spaces after commas in a list if there are none, wrap long lines, etc.)
- [x] Check image and plot sizes. If possible, making them responsive is highly recommended.
- [x] Check the appearance on a smartphone (check Google to see how to adapt the appearance of your browser to display pages as if they were seen from a smartphone, this is usually done via the web developer tools). This is not a top priority for all examples, but if there are a few easy and straightforward changes to make that can improve the experience, let’s do it.
- [x] Check the updated notebook with the original notebook

## Visual appearance - Gallery

- [x] Check the thumbnail is visually appealing
- [x] Check the project title is well formatted (e.g. `Ml Annotators` to `ML Annotators`), if not, add/update the `examples_config.title` field in `anaconda-project.yml`
- [x] Check the project description is appropriate, if not, update the `description` field in `anaconda-project.yml`

## Workflow (after you have made the changes above)

- [x] Run successfully `doit validate:<projectname>`
- [x] Run successfully `doit test:<projectname>`
- [x] Run successfully `doit doc_one –name <projectname>`. It’s better if the project notebook(s) is saved with its outputs (but be sure to clear outputs before committing to the examples repo!) when building the docs. Then open this file in your browser `./builtdocs/index.html` and check how the site looks.
- [x] If you’re happy with all the above, open a PR. Reminder, clear notebook outputs before pushing to the PR.

### CODE REVIEW

File: gapminders/gapminders.ipynb [hunk 1]
```diff
@@ -11,16 +11,9 @@
...
    "source": [
-    "<img src=\"https://panel.holoviz.org/_static/logo_stacked.png\" width=150 align=\"left\" margin=20px>"
-   ]
-  },
-  {
-   "cell_type": "markdown",
-   "metadata": {},
-   "source": [
-    "The [Panel](https://panel.holoviz.org) library from [HoloViz](https://holoviz.org) lets you make widget-controlled apps and dashboards from a wide variety of [plotting libraries and data types](https://panel.holoviz.org/reference/index.html#panes).  \n",
+    "In his Gapminder example during thes 2006 TED Talk, Hans Rosling debunked stereotypes about developed and undeveloped countries using statistics and data visualization, revealing the nuanced reality of global development. We will be recreating this example using four different plotting libraries which will be controlled by widgets from Panel.\n",
```
Comment (ahuang11) [inline gapminders/gapminders.ipynb:14]:
```suggestion
    "In this Gapminder example during thes 2006 TED Talk, Hans Rosling debunked stereotypes about developed and undeveloped countries using statistics and data visualization, revealing the nuanced reality of global development. We will be recreating this example using four different plotting libraries which will be controlled by widgets from Panel.\n",
```

File: gapminders/gapminders.ipynb [hunk 2]
```diff
@@ -39,17 +30,18 @@
...
+    "pn.extension('vega', 'plotly', defer_load=True, sizing_mode=\"stretch_width\")\n",
+    "mpl.use('agg')"
    ]
   },
   {
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "We need to define some configuration"
+    "We'll begin by defining some global variables for our plots."
```
Comment (ahuang11) [inline gapminders/gapminders.ipynb:44]:
```suggestion
    "We'll begin by defining some constant variables for our plots."
```

File: gapminders/gapminders.ipynb [hunk 3]
```diff
@@ -99,6 +90,13 @@
     "dataset.sample(10)"
    ]
   },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We'll also create a global variable `YEARS` containing all the unique years in our dataset."
```
Comment (ahuang11) [inline gapminders/gapminders.ipynb:97]:
```suggestion
    "We'll also create a constant variable `YEARS` containing all the unique years in our dataset."
```

**Comment by ahuang11** [comment]
Under Files changed, you may use `Add suggestions to batch` and then commit them together https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes

File: gapminders/anaconda-project.yml
Comment (Azaya89) [inline gapminders/anaconda-project.yml:1]: I think you should also update the lower pins of the other packages as well, not just `python` and `hvplot`.

File: gapminders/gapminders.ipynb [hunk 4]
```diff
@@ -365,19 +355,19 @@
...
+    "logo = pn.pane.PNG(\n",
     "    \"https://panel.holoviz.org/_static/logo_stacked.png\",\n",
     "    link_url=\"https://panel.holoviz.org\", embed=False, width=150, align=\"center\"\n",
     ")\n",
-    "    \n",
-    "desc = \"\"\"## 🎓 Info\n",
+    "\n",
+    "desc = pn.pane.Markdown(\"\"\"## 🎓 Info\n",
     "\n",
     "The [Panel](http://panel.holoviz.org) library from [HoloViz](https://holoviz.org)\n",
```
Comment (Azaya89) [inline gapminders/gapminders.ipynb:365]:
```suggestion
    "The [Panel](https://panel.holoviz.org) library from [HoloViz](https://holoviz.org)\n",
```

File: gapminders/gapminders.ipynb [hunk 5]
```diff
@@ -415,7 +436,7 @@
    "source": [
     "## Configure the template\n",
     "\n",
-    "Let us layout out the app in the nicely styled [FastListTemplate](https://panel.holoviz.org/reference/templates/FastListTemplate.html)."
+    "Let us layout out the app in the nicely styled with [FastListTemplate](https://panel.holoviz.org/reference/templates/FastListTemplate.html)."
```
Comment (Azaya89) [inline gapminders/gapminders.ipynb:439]: The grammar here doesn't sound right, I think.
```suggestion
    "Let us layout out the app in the nicely styled [FastListTemplate](https://panel.holoviz.org/reference/templates/FastListTemplate.html)."
```

File: gapminders/gapminders.ipynb (hunk 4 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb:365]: Thanks, I will make this change.

File: gapminders/gapminders.ipynb (hunk 5 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb:439]: Thanks, I will make this change.

**Comment by jtao1** [comment]
gapminders with cache
gapminders no cache

**Comment by droumis** [comment]
cool, let's not use caching unless there's a clear benefit

File: gapminders/anaconda-project.yml [hunk 6]
```diff
@@ -10,24 +10,23 @@ examples_config:
...
-- plotly
-- holoviews
-- hvplot
-- panel >=1
-- param
-- bokeh >=3
-- altair
-- pandas
+- python=3.11
+- notebook>=7.0.8
```
Comment (maximlt) [inline gapminders/anaconda-project.yml:21]: Please re-lock after that.

```suggestion
- notebook<7
```

File: gapminders/anaconda-project.yml [hunk 7]
```diff
@@ -39,7 +38,12 @@ commands:
     notebook: gapminders.ipynb
 
 variables: {}
-downloads: {}
+downloads: 
+  gapminders_data:
+    url: 'https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv'
```
Comment (maximlt) [inline gapminders/anaconda-project.yml:43]: Let's update this URL not to point to the `master` branch but to a specific commit, so it's more stable.

File: gapminders/gapminders.ipynb [hunk 8]
```diff
@@ -83,20 +83,18 @@
    "metadata": {},
    "outputs": [],
    "source": [
-    "@pn.cache\n",
     "def get_dataset():\n",
-    "    url = 'https://raw.githubusercontent.com/plotly/datasets/master/gapminderDataFiveYear.csv'\n",
-    "    return pd.read_csv(url)"
+    "    return pd.read_csv('data/gapminders.csv')\n",
```
Comment (maximlt) [inline gapminders/gapminders.ipynb:87]: The `get_dataset` function was useful when decorated with `pn.cache` and fetching data remotely. I think it's no longer required, you could just replace it with `dataset = d.read_csv('data/gapminders.csv')`.

File: gapminders/anaconda-project.yml (hunk 6 above)
Comment (jtao1) [reply gapminders/anaconda-project.yml:21]: Changed

File: gapminders/gapminders.ipynb (hunk 8 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb:87]: I see, that makes sense.

File: test_data/gapminders/gapminder_test_data.csv [hunk 9]
```diff
@@ -0,0 +1,151 @@
+,country,year,pop,continent,lifeExp,gdpPercap
```
[the remark this answers is not in the document]
Comment (jtao1) [reply test_data/gapminders/gapminder_test_data.csv:1]: Ok I will make that change.

File: gapminders/gapminders.ipynb (hunk 8 above)
Comment (droumis) [reply gapminders/gapminders.ipynb]: @jtao1 Not only is this link broken, but we are also not using cache in this cell below this anymore. Checking links was one of the items in the checklist on this PR description. Please review the checklist again in greater detail. When you have confirmed that you have completed all the steps, message me to continue with the review. 

image

**Comment by droumis** [comment]
@jtao1  Please press 'resolve conversation' on any comments that have been completed

File: gapminders/gapminders.ipynb (hunk 8 above)
Comment (droumis) [reply gapminders/gapminders.ipynb]: there's also still a pn.cache decorator elsewhere that should be removed

**Comment by maximlt** [comment]
> @jtao1 Please press 'resolve conversation' on any comments that have been completed

I think I asked Jason and Isaiah not to resolve the conversations themselves, or at least not until they have committed the changes, since an example review could entail *many* of these conversations and I wanted to be able to review their changes and maybe iterate on them.

**Comment by droumis** [comment]
Yep, the relevant changes should be committed before they mark something as resolved

File: gapminders/anaconda-project.yml (hunk 7 above)
Comment (jtao1) [reply gapminders/anaconda-project.yml:43]: I found the plotly dataset and replaced `master` with the commit id.

File: gapminders/anaconda-project.yml
Comment (jtao1) [reply gapminders/anaconda-project.yml:1]: This has been updated to include the lower pins.

File: gapminders/gapminders.ipynb (hunk 4 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb]: I've made a PR with this change.

File: gapminders/gapminders.ipynb (hunk 5 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb]: I've made a PR with this change.

File: test_data/gapminders/gapminder_test_data.csv (hunk 9 above)
[the remark this answers is not in the document]
Comment (jtao1) [reply test_data/gapminders/gapminder_test_data.csv]: I've made a commit to remove that line from the test_data.

File: gapminders/gapminders.ipynb (hunk 8 above)
Comment (jtao1) [reply gapminders/gapminders.ipynb]: I've made a commit to remove that link since it no longer is useful and I've removed the pn.cache decorator.

**Comment by droumis** [comment]
ok I updated some text and headings and also added the settings widgets to the notebook version of the app. going to merge now

**Review by droumis — APPROVED** [review]

**Pull request merged**